[Experimental] Further relax general deletion of script-written website data to 400 days

sysrqb opened the pull request:
#### 772ed3bc762f49c57fd4349fb0f4feac6e62057e
<pre>
[Experimental] Further relax general deletion of script-written website data to 400 days
<a href="https://bugs.webkit.org/show_bug.cgi?id=293793">https://bugs.webkit.org/show_bug.cgi?id=293793</a>
<a href="https://rdar.apple.com/152095900">rdar://152095900</a>

Reviewed by NOBODY (OOPS!).

274398@main introduced a &quot;long&quot; DataRemovalFrequency that was set at 30
operational days. This patch extends that lifetime to 400 operational days.
This roughly aligns web storage with the proposed upper-limit on cookie age [0]
(although the cookie age is strictly based on calendar days). We may change
this behavior in the future, again.

This patch fixes some flakey tests by giving
ResourceLoadStatisticsStore::processStatisticsAndDataRecords a completion
handler that is called when the process completes. It seems like our thread
hopping performance has regressed a bit since these tests were written. This
patch also adds a new test that verifies the state of the website data and
database after 399 days.

The code changes make explicit how each time period is used:
  - deletion of prevalent domain data (30 operation days)
  - deletion of script-written website data when link decoration filtering is disabled (7 operational days)
  - deletion of script-written website data when link decoration filtering is enabled (400 operational days)

[0] <a href="https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-cookie-lifetime-limits">https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-cookie-lifetime-limits</a>

* LayoutTests/http/tests/resourceLoadStatistics/operating-dates-all-website-data-removed.html:
* LayoutTests/http/tests/resourceLoadStatistics/prevalent-resource-without-user-interaction.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-long-deletion-expected.txt: Copied from LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-long-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration.html:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::processStatisticsAndDataRecords):
(WebKit::ResourceLoadStatisticsStore::scheduleStatisticsProcessingRequestIfNecessary):
(WebKit::ResourceLoadStatisticsStore::resetParametersToDefaultValues):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccessInternal):
(WebKit::ResourceLoadStatisticsStore::logUserInteraction):
(WebKit::ResourceLoadStatisticsStore::areAllUnpartitionedThirdPartyCookiesBlockedUnder):
(WebKit::ResourceLoadStatisticsStore::hasHadRecentWebPushInteraction const):
(WebKit::ResourceLoadStatisticsStore::hasHadUnexpiredRecentUserInteraction):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllWebsiteDataFor):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllButCookiesFor):
(WebKit::ResourceLoadStatisticsStore::updateOperatingDatesParameters):
(WebKit::ResourceLoadStatisticsStore::includeTodayAsOperatingDateIfNecessary):
(WebKit::ResourceLoadStatisticsStore::hasStatisticsExpired const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::scheduleStatisticsAndDataRecordsProcessing):
(WebKit::WebResourceLoadStatisticsStore::resourceLoadStatisticsUpdated):
(WebKit::WebResourceLoadStatisticsStore::hasHadUserInteraction):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/772ed3bc762f49c57fd4349fb0f4feac6e62057e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112875 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58203 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28031 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35933 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81752 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110597 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22218 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97034 "Found 1 new API test failure: TestWebKitAPI.WebPushDTest.NotificationClickExtendsITPCleanupTimerBy30Days (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62124 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21652 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15174 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57637 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91590 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15207 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115994 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34740 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25646 "Found 1 new test failure: fast/webgpu/nocrash/fuzz-291180.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90786 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35117 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90534 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35449 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13221 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30496 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34645 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40201 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34391 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37752 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -440,14 +442,14 @@ void ResourceLoadStatisticsStore::removeDataRecords(CompletionHandler<void()>&&
+    removeDataRecords([weakThis = WeakPtr { *this }, completionHandler = WTFMove(completionHandler)] () mutable {
         ASSERT(!RunLoop::isMain());
         RefPtr protectedThis = weakThis.get();
         if (!protectedThis)
```
You're not calling the completion handler here

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -265,9 +265,12 @@ void WebResourceLoadStatisticsStore::scheduleStatisticsAndDataRecordsProcessing(
+        if (RefPtr statisticsStore = m_statisticsStore) {
+            statisticsStore->processStatisticsAndDataRecords([weakThis = ThreadSafeWeakPtr { *this }, completionHandler = WTFMove(completionHandler)] () mutable {
+                if (RefPtr store = weakThis.get())
+                    store->postTaskReply(WTFMove(completionHandler));
```
Same thing here, if `this` is gone the completion handler will never be called

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
```diff
@@ -440,14 +442,14 @@ void ResourceLoadStatisticsStore::removeDataRecords(CompletionHandler<void()>&&
 }
 
-void ResourceLoadStatisticsStore::processStatisticsAndDataRecords()
+void ResourceLoadStatisticsStore::processStatisticsAndDataRecords(CompletionHandler<void()>&& completionHandler)
```
Do you need to add this completion handler in this change? It's not clear to me why this is related to extending the days on script-written data deletion.

sysrqb replied:
The tests are flakey. I'll add this into the commit message. If we don't move the completion handler, then many of the resourceLoadStatistics tests will call `processStatisticsAndDataRecords` multiple times, and those calls will overlap. This invalidates some of the test results because the test expectations depend on the state of the database at specific times. The current code runs processStatisticsAndDataRecords asynchronously, so there's no way to verify the state of website data at specific time points.

Commit:
[Experimental] Further relax general deletion of script-written website data to 400 days
https://bugs.webkit.org/show_bug.cgi?id=293793
rdar://152095900

Reviewed by NOBODY (OOPS!).

274398@main introduced a "long" DataRemovalFrequency that was set at 30
operational days. This patch extends that lifetime to 400 operational days.
This roughly aligns web storage with the proposed upper-limit on cookie age [0]
(although the cookie age is strictly based on calendar days). We may change
this behavior in the future, again.

This patch fixes some flakey tests by giving
ResourceLoadStatisticsStore::processStatisticsAndDataRecords a completion
handler that is called when the process completes. It seems like our thread
hopping performance has regressed a bit since these tests were written. This
patch also adds a new test that verifies the state of the website data and
database after 399 days.

The code changes make explicit how each time period is used:
  - deletion of prevalent domain data (30 operation days)
  - deletion of script-written website data when link decoration filtering is disabled (7 operational days)
  - deletion of script-written website data when link decoration filtering is enabled (400 operational days)

[0] https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-cookie-lifetime-limits

* LayoutTests/http/tests/resourceLoadStatistics/operating-dates-all-website-data-removed.html:
* LayoutTests/http/tests/resourceLoadStatistics/prevalent-resource-without-user-interaction.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-long-deletion-expected.txt: Copied from LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-before-long-deletion.html: Copied from LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion.html.
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt:
* LayoutTests/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration.html:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-with-filtered-link-decoration-after-long-deletion-expected.txt:
* LayoutTests/platform/glib/http/tests/resourceLoadStatistics/website-data-removal-for-site-navigated-to-without-link-decoration-expected.txt:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::processStatisticsAndDataRecords):
(WebKit::ResourceLoadStatisticsStore::scheduleStatisticsProcessingRequestIfNecessary):
(WebKit::ResourceLoadStatisticsStore::resetParametersToDefaultValues):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccess):
(WebKit::ResourceLoadStatisticsStore::grantStorageAccessInternal):
(WebKit::ResourceLoadStatisticsStore::logUserInteraction):
(WebKit::ResourceLoadStatisticsStore::areAllUnpartitionedThirdPartyCookiesBlockedUnder):
(WebKit::ResourceLoadStatisticsStore::hasHadRecentWebPushInteraction const):
(WebKit::ResourceLoadStatisticsStore::hasHadUnexpiredRecentUserInteraction):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllWebsiteDataFor):
(WebKit::ResourceLoadStatisticsStore::shouldRemoveAllButCookiesFor):
(WebKit::ResourceLoadStatisticsStore::updateOperatingDatesParameters):
(WebKit::ResourceLoadStatisticsStore::includeTodayAsOperatingDateIfNecessary):
(WebKit::ResourceLoadStatisticsStore::hasStatisticsExpired const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::scheduleStatisticsAndDataRecordsProcessing):
(WebKit::WebResourceLoadStatisticsStore::resourceLoadStatisticsUpdated):
(WebKit::WebResourceLoadStatisticsStore::hasHadUserInteraction):
